server/app/services/apikey/ApiKeyService.java

```diff
@@ -0,0 +1,302 @@
+import java.util.regex.Pattern;
+import javax.crypto.KeyGenerator;
+import models.ApiKey;
+import org.apache.commons.net.util.SubnetUtils;
+import play.Environment;
+import play.data.DynamicForm;
+import repository.ApiKeyRepository;
+import services.DateConverter;
+import services.program.ProgramNotFoundException;
+import services.program.ProgramService;
+
+/**
+ * Service management of the resource backed by {@link ApiKey}.
+ *
+ * <p>CiviForm uses API keys to implement HTTP basic auth
+ * (https://en.wikipedia.org/wiki/Basic_access_authentication) for authenticating API requests, as
+ * well as for authorizing requests based on the permission grants stored with the key in the
+ * database.
+ *
+ * <p>For a given key, CiviForm stores the key secret in plain text along with the key secret hashed
+ * with a secret salt value provided to the server as a configuration variable.
+ *
+ * <p>The plaintext secret is revealed to the admin user once after creation, after which it is not
+ * recoverable from CiviForm.
+ */
+public class ApiKeyService {
+
+  public static final String FORM_FIELD_NAME_KEY_NAME = "keyName";
+  public static final String FORM_FIELD_NAME_EXPIRATION = "expiration";
+  public static final String FORM_FIELD_NAME_SUBNET = "subnet";
+  public static final String DEFAULT_API_KEY_SECRET_SALT = "changeme";
+
+  private static final int KEY_ID_LENGTH = 128;
+  private static final int KEY_SECRET_LENGTH = 256;
+
+  private final ApiKeyRepository repository;
+  private final Environment environment;
+  private final ProgramService programService;
+  private final DateConverter dateConverter;
+  private final String secretSalt;
+  private final boolean banGlobalSubnet;
+
+  @Inject
+  public ApiKeyService(
+      ApiKeyRepository repository,
+      Environment environment,
+      ProgramService programService,
+      DateConverter dateConverter,
+      Config config) {
+    this.repository = checkNotNull(repository);
+    this.environment = checkNotNull(environment);
+    this.programService = checkNotNull(programService);
+    this.dateConverter = checkNotNull(dateConverter);
+    this.secretSalt = checkNotNull(config).getString("api_secret_salt");
+    this.banGlobalSubnet = checkNotNull(config).getBoolean("api_keys_ban_global_subnet");
+  }
+
+  /**
+   * Creates a new {@link ApiKey} with the data in {@code form} if it passes validation.
+   *
+   * @param form Stores data specified by the user for creation of the key, as well as validation
+   *     errors
+   * @param profile The user profile of the account creating this API key, used for recording who
+   *     created it
+   * @return a result object containing the credentials and created key if successful, form with
+   *     validation messages if not
+   */
+  public ApiKeyCreationResult createApiKey(DynamicForm form, CiviFormProfile profile)
+      throws ProgramNotFoundException {
+    if (environment.isProd() && secretSalt.equals(DEFAULT_API_KEY_SECRET_SALT)) {
```

Comment:
Shipping a in-code default that can't be used is a bit smelly.

Do we have dev configs that we could put that in instead, and then check that the salt is set at all here?

Reply:
> Do we have dev configs that we could put that in instead, and then check that the salt is set at all here?

We could, but the result is that we've have to require the secret salt to be present for the server to start at all, rather than this behavior which will just block this one feature if it hasn't been set yet.

I think I prefer the option that doesn't risk downtime.